src/stream.satyg

```diff
@@ -3,30 +3,63 @@
 @import: option-ext
 @import: list-ext
 @import: tuple
-
+@import: result
 @import: string
-@import: debug
+@import: array
 
 module Stream : sig
   type 'a t
-  val cons : 'a thunk -> ('a t) thunk -> 'a t
   val nil : unit -> 'a t
+  val cons : 'a thunk -> ('a t) thunk -> 'a t
+  val unfold : ('a -> ('b * 'a) option) -> 'a -> 'b t
+  val const : 'a -> 'a t
+  val nat : int t
+  val fib : int t
   val uncons : 'a t -> ('a * 'a t) option
+  val null : 'a t -> bool
   val head : 'a t -> 'a option
   val tail : 'a t -> ('a t) option
+  val length : 'a t -> int
+  val last : 'a t -> 'a option
+  val nth : int -> 'a t -> 'a option
+  val step-by : int -> 'a t -> 'a t
+  val append : 'a t -> 'a t -> 'a t
+  val chain : 'a t -> 'a t -> 'a t
+  val zip : 'a t -> 'b t -> ('a * 'b) t
   val map : ('a -> 'b) -> 'a t -> 'b t
+  val iter : ('a -> unit) -> 'a t -> unit
   val filter : ('a -> bool) -> 'a t -> 'a t
-  val of-list : 'a list -> 'a t
-  val to-list : 'a t -> 'a list
-  val unfold : ('a -> ('b * 'a) option) -> 'a -> 'b t
-  val fold : ('a -> 'b -> 'a) -> 'a -> 'b t -> 'a
-  val take : int -> 'a t -> 'a t
+  val filter-map : ('a -> 'b option) -> 'a t -> 'b t
+  val enumerate : 'a t -> (int * 'a) t
+  val drop-while : ('a -> bool) -> 'a t -> 'a t
+  val take-while : ('a -> bool) -> 'a t -> 'a t
+  val map-while : ('a -> 'b option) -> 'a t -> 'b t
   val drop : int -> 'a t -> 'a t
-  val zip : 'a t -> 'b t -> ('a * 'b) t
+  val take : int -> 'a t -> 'a t
+  val flat-map : ('a -> 'b t) -> 'a t -> 'b t
+  val flatten : ('a t) t -> 'a t
+  val tap : ('a -> unit) -> 'a t -> 'a t
+  val partition : ('a -> bool) -> 'a t -> ('a t * 'a t)
+  val fold-option : 'b -> ('a -> 'b -> 'b option) -> 'a t -> 'b option
+  val fold-result : 'b -> ('a -> 'b -> 'b 'e result) -> 'a t -> 'b 'e result
+  val iter-option : ('a -> unit option) -> 'a t -> unit option
+  val iter-result : ('a -> unit 'e result) -> 'a t -> unit 'e result
+  val all : ('a -> bool) -> 'a t -> bool
+  val any : ('a -> bool) -> 'a t -> bool
+  val find : ('a -> bool) -> 'a t -> 'a option
+  val find-map : ('a -> 'b option) -> 'a t -> 'b option
+  val find-result : ('a -> bool 'e result) -> 'a t -> ('a option) 'e result
+  val position : ('a -> bool) -> 'a t -> int option
   val unzip : ('a * 'b) t -> ('a t * 'b t)
-  val fib : int t
-  val nat : int t
-  val const : 'a -> 'a t
+  val of-list : 'a list -> 'a t
+  val of-string : string -> Char.t t
+  val of-array : 'a Array.t -> 'a t
+  val of-option : 'a option -> 'a t
+  val to-list : 'a t -> 'a list
+  val to-array : 'a t -> 'a Array.t
+
+  % deprecated
+  val fold : ('b -> 'a -> 'b) -> 'b -> 'a t -> 'b  % TODO: swap arguments
```

Comment:
Is `fold` just deprecated, not renamed?

Reply:
I wanted to name the improved version `fold`, too.
My plan is deprecate the current version for the moment and swap the argument in version 2.0.0